galpy/df/quasiisothermaldf.py

```diff
@@ -1685,6 +1688,194 @@ def sampleV(self,R,z,n=1):
+
+            z - array of height (can be Quantity)
+            
+            R_pixel, z_pixel= the pixel size for creating the grid for
+                   interpolation (in natural unit)
+            
+            num_std= number of standard deviation to be considered outliers
+                      sampled separately from interpolation
+                      
+            R_min, R_max, z_max= optional edges of the grid
+    
+        OUTPUT:
+            
+            coord_v= a numpy array containing the sampled velocity, (vR, vT, vz),
+                     where each row correspond to the row of (R,z)
+    
+        HISTORY:
+            
+            2018-08-10 - Written - Samuel Wong (University of Toronto)
+            
+        """
+        #Initialize output array
+        coord_v= numpy.empty((numpy.size(R), 3))
+        #Since the sign of z doesn't matter, work with absolute value of z
+        z= numpy.abs(z)
+        #Separate the coodinates into outliers and normal points.
+        mean_R= numpy.mean(R)
+        std_R= numpy.std(R)
+        mean_z= numpy.mean(z)
+        std_z= numpy.std(z)
+        #Define outliers as points either more than <num_std> std away from
+        #mean, or those outside user-specified grid
+        if R_min is None:
+            R_min_temp= 0
+        else:
+            R_min_temp= R_min
+        if R_max is None:
+            R_max_temp= numpy.max(R)
+        else:
+            R_max_temp= R_max
+        if z_max is None:
+            z_max_temp= numpy.max(z)
+        else:
+            z_max_temp= z_max
+        mask= numpy.any([numpy.abs(R - mean_R) > num_std*std_R, 
+                       numpy.abs(z - mean_z) > num_std*std_z,
+                       R < R_min_temp, R > R_max_temp, z > z_max_temp],axis = 0)
+        outliers_R= R[mask]
+        outliers_z= z[mask]
+        normal_R= R[~mask]
+        normal_z= z[~mask]
+        #Sample the velocity of outliers directly (without interpolation)
+        outlier_coord_v= numpy.empty((outliers_R.size, 3))
+        for i in range(outliers_R.size):
+            outlier_coord_v[i]= self.sampleV(outliers_R[i], outliers_z[i])[0]
+        #Prepare for optimizing maxVT on a grid
+        #Edges of grid determined by input, unless prespecified by users
+        if R_min is None:
+            R_min= numpy.min(normal_R)
+        if R_max is None:
+            R_max= numpy.max(normal_R)
+        if z_max is None:
+            z_max= numpy.max(normal_z)
+        z_min= 0. #Always start grid at z=0 for stars close to plane
+        #Get the new hash of the parameters of grid
+        new_hash= hashlib.md5(numpy.array([R_min,R_max,z_max,R_pixel,z_pixel])).hexdigest()
+        #Reuse old interpolated object if new hash matches the old one
+        if new_hash == self._maxVT_hash:
+            ip_max_vT= self._maxVT_ip
+        #Generate a new interpolation object if different from before
+        else:
+            R_number= int((R_max - R_min)/R_pixel)
```

Comment:
Should probably first do ``floor`` or ``ceil`` here, before turning into an int (sorry for not spotting before).

Reply:
That's true. `ceil` is better since `kx` and `ky` will be less likely to become zero (and give error) in cases where the pixels are too large.

Reply:
@jobovy Speaking of that, should I raise an error if pixel is too large that kx or ky become zero?

Reply:
``R_number`` should always be at least 2, no? Even if ``R_max-R_min`` = ``R_pixel``, no?